firedrake/extrusion_numbering.pyx

```diff
@@ -0,0 +1,1015 @@
+            for i in range(top - bottom - 1):
+                top_masks[foffset + i] = ctop[coffset + i]
+                bottom_masks[foffset + i] = cbottom[coffset + i]
+            facet += 1
+    bottom = PETSc.IS().createGeneral(bottom_masks, comm=PETSc.COMM_SELF)
+    top = PETSc.IS().createGeneral(top_masks, comm=PETSc.COMM_SELF)
+    return op2.ExtrudedSet.EntityMask(section=section, bottom=bottom, top=top)
+
+
+def interior_facet_entity_masks(mesh):
+    """Compute a pair of masking integers for each interior facet in the
+    extruded mesh.
+
+    This integer indicates for each facet, which topological entities
+    in the closure of the support of the facet are on the boundary of
+    the domain.  If the ith bit in the integer is on, that indicates
+    that the ith entity is on the boundary, meaning that the
+    appropriate boundary mask should be used to discard element tensor
+    contributions when assembling bilinear forms.
+
+    :arg mesh: the extruded mesh.
+    :returns: a tuple of section, bottom, and top masks.  The section
+        records the number of entities in each column and the offset
+        in the masking arrays for the start of each column.
+    """
+    label = "interior_facets"
+    dm = mesh._plex
+
+    pStart, pEnd = dm.getChart()
+
+    cell_numbering = mesh._cell_numbering
+    renumbering = mesh._plex_renumbering.indices
+    fStart, fEnd = dm.getHeightStratum(1)
+
+    layers = mesh.interior_facets.set.layers_array
+    top_masks = numpy.zeros((numpy.sum(layers[:, 1] - layers[:, 0] - 1), 2),
+                            dtype=IntType)
+    bottom_masks = numpy.zeros_like(top_masks)
+
+    section = PETSc.Section().create(comm=PETSc.COMM_SELF)
+    section.setChart(0, mesh.exterior_facets.set.total_size)
+    for p in range(*section.getChart()):
+        section.setDof(p, layers[p, 1] - layers[p, 0] - 1)
+    section.setUp()
+
+    csection, cbottom, ctop = mesh.cell_set.masks
+    facet = 0
+
+    for p in range(pStart, pEnd):
+        point = renumbering[p]
+        if fStart <= point < fEnd:
+            if dm.getLabelValue(label, point) == -1:
+                continue
+            bottom = layer_extents[point, 0]
+            top = layer_extents[point, 1]
+            for j, c in enumerate(dm.getSupport(point)):
+                cell = cell_numbering.getOffset(c)
+                cbottom = layer_extents[c, 0]
+                coffset = csection.getOffset(cell)
+                foffset = section.getOffset(facet)
+                for i in range(top - bottom - 1):
+                    top_masks[foffset + i, j] = ctop[coffset + i + bottom - cbottom]
+                    bottom_masks[foffset + i, j] = cbottom[coffset + i + bottom - cbottom]
+            facet += 1
+    bottom = PETSc.IS().createGeneral(bottom_masks, comm=PETSc.COMM_SELF)
+    top = PETSc.IS().createGeneral(top_masks, comm=PETSc.COMM_SELF)
+    return op2.ExtrudedSet.EntityMask(section=section, bottom=bottom, top=top)
```

Comment:
Problem here.  By the time we get in to PyOP2, we don't know we've got a map that's come from a concatenated thing. So how do we deal with that.  Do we treat the two cells as a "macro cell" in some sense, and glue together the masks for the interior facets.  Or do we actually remember this information.